### PR TITLE
Encoding package tarball url for scoped packages

### PR DESCRIPTION
**Pull request by hugoaponte**
We have our private couchdb registry to keep our scoped packages but also continuously syncing with the official registry.

I hit this issue: https://github.com/npm/npm/issues/8770

I worked in a fix and committed it to this pull request.

Please bear with me as this is my first time contributing to npm. I tried to validate the tests, but several tests were broken even before my changes.

Please share any feedback.

### CODE REVIEW

**Comment by iarna** [comment]
Thank you for this! This does need a test to reproduce the bug found in #8770.

(CI passes for this branch https://travis-ci.org/npm/npm/builds/98592627 so other than that this looks good.)

**Comment by hapontemloreto** [comment]
Hi Rebecca

I put in place a couple of tests to cover this bug. Builds and tests are succeeding. It is my first time with tap test and npm tests, could you take a look?

Thanks in advance,

Hugo

**Comment by hapontemloreto** [comment]
Hi Rebecca

Kindly ping. Thanks.

Hugo

File: test/tap/enconding-tar-ball-for-scoped-packages.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        tarball: 'https://registry.npmjs.org/@scope/xyzzy-wibble/-/@scope/xyzzy-wibble-1.8.0.tgz'
+      }
+    }
+  }
+}
+
+var lastTarball
+
+test('setup', function (t) {
+  npm.load(function (a, b) {
```
Comment (iarna) [inline test/tap/enconding-tar-ball-for-scoped-packages.js]: This and the next line can be:
```
npm.load({loglevel: 'silly'}, function (a, b) {
````

Comment (hapontemloreto) [inline test/tap/enconding-tar-ball-for-scoped-packages.js]: Thanks. Done.

Comment (hapontemloreto) [inline test/tap/enconding-tar-ball-for-scoped-packages.js]: Hi

Thanks so much for the feedback. It is done.

Hugo

**Comment by hapontemloreto** [comment]
Hi

Thanks so much for the feedback. It is done.

Hugo

**Comment by iarna** [comment]
Test looks great! I'm scheduling this for this week.

**Comment by iarna** [comment]
So when I was writing the changelog entry for this I became concerned.

I'd been looking at these changes in the small, but now that I'm looking at the bigger picture I'm no longer convinced that this is something we can accept. 

Let me see if I understand correctly– I might be wrong here:

You have a private registry that returns distribution URLs that can't be fetched as is– they require some of the URL that they return to be escaped? Is that correct?

Because if that's the case, that's _definitely_ a bug in your registry and should be fixed there. `npm` should NOT be mangling the URLs it gets back from registries, it should be handling them as-is.

**Comment by iarna** [comment]
The reason I would insist that you fix the bug in your registry, rather than working around it here, is that this potentially could break other third party registries that return it unescaped on purpose. We support all registry products, so it needs to not just break the `npm` registry but also the many varied third party ones. (There are several open source ones and several proprietary ones.)

**Comment by hapontemloreto** [comment]
Thanks for the feedback. I will review in detail. I am not sure the issue is on the registry. I am using the package https://github.com/npm/npm-registry-couchapp, which is actually being deprecated.

I see what you meant about the distribution URL. I am not sure the json that goes under .versions[X.Y.Z] in the registry is created by npm when publishing the package or by the registry when receiving it. Any hint on that would be appreciated.

I will take a look again this week.

Thanks

Hugo

**Comment by iarna** [comment]
I'm definitely surprised that `npm-registry-couchapp` would be doing that, and despite it's kind of deprecated status, if you can find a reproducer for this I'd be happy to work with you in fixing this over there.

**Comment by a-hegerath** [comment]
Sorry for jumping into this discussion, but I recently ran into this issue, too, and I am not sure if it is really only a bug in the registry. Saying that, I should add that I am quite new to npm so I might be wrong.

I got the error with the wrong URLs for scoped packages in private registries with a different registry (Sonatype Nexus). As far as I can see, for the scoped package "@angular/compiler" npm generates a URL that ends with "@angular%2fcompiler":

>npm install --verbose --always-auth true
npm info it worked if it ends with ok
npm verb cli [ 'D:\\ah\\nodejs\\node.exe',
npm verb cli   'D:\\ah\\nodejs\\node_modules\\npm\\bin\\npm-cli.js',
npm verb cli   'install',
npm verb cli   '--verbose',
npm verb cli   '--always-auth',
npm verb cli   'true' ]
npm info using npm@3.9.5
npm info using node@v6.2.2
npm verb mapToRegistry no registry URL found in name for scope @angular
npm verb mapToRegistry no registry URL found in name for scope @angular
npm verb mapToRegistry no registry URL found in name for scope @angular
npm verb mapToRegistry no registry URL found in name for scope @angular
npm verb mapToRegistry no registry URL found in name for scope @angular
npm verb mapToRegistry no registry URL found in name for scope @angular
npm verb request uri https://<host>/nexus/content/groups/npm_public_group/@angular%2fcompiler

While the public registry deals with this fine, my private registry returns for a GET request some package information with invalid tarball URLs, which leads to the 404 error. This seems like a bug in the registry. But for a GET with the proper URL (that ends with '@angular/compiler' instead), the returned package information contains valid tarball URLs. 

To me the question is, why does npm not apply URL encoding for the slash in the scoped package name in any case? Would this break compatibility or raise some other issues?

Thanks,
Andre

**Comment by iarna** [comment]
@a-hegerath It looks to me like that isn't your problem. Rather something is off about your registry URL because:

```
https:///nexus/content/groups/npm_public_group/@angular%2fcompiler
```
Is not a valid URL. There's no hostname there... it should be `https://HOSTNAME/path/path/path`.

> To me the question is, why does npm not apply URL encoding for the slash in the scoped package name in any case? Would this break compatibility or raise some other issues?

As I said before, if YOUR registry told us that it shouldn't be encoded, we aren't going to fiddle with that. The registry you're using needs to return URLs that can be fetched. How tarball paths are built is entirely up to the registry hosting them and npm has no insight into them. As far as npm is concerned those are opaque strings with no meaning at all. They could be shasums without package names in them for all npm knows.

**Comment by iarna** [comment]
(wrong button =D)

**Comment by iarna** [comment]
We're closing this pull request as it has gone sixty days without activity.  In our experience pull requests that go this long are unlikely to land.  If you have updates to it, please feel free to open a new pull request.

For more information about our new issue aging policies and why we've instituted them please see our [blog post](http://blog.npmjs.org/post/161832149430/npm-the-npm-github-issue-tracker-and-you).

**Pull request closed without merging**